datafusion/expr/src/expr_schema.rs

```diff
@@ -290,13 +293,30 @@ impl ExprSchemable for Expr {
     /// This function errors when it is impossible to cast the
     /// expression to the target [arrow::datatypes::DataType].
     fn cast_to<S: ExprSchema>(self, cast_to_type: &DataType, schema: &S) -> Result<Expr> {
+        let this_type = self.get_type(schema)?;
+        if this_type == *cast_to_type {
+            return Ok(self);
+        }
+
         // TODO(kszucs): most of the operations do not validate the type correctness
         // like all of the binary expressions below. Perhaps Expr should track the
         // type of the expression?
-        let this_type = self.get_type(schema)?;
-        if this_type == *cast_to_type {
-            Ok(self)
-        } else if can_cast_types(&this_type, cast_to_type) {
+
+        // TODO(jackwener): Handle subqueries separately, need to refactor it.
```

Comment:
Is this worth tracking with a ticket? I am not sure if  this comment mean you want to improve the code or if there is some bug / limitiation

Reply:
add a ticket #5877. 

>  I am not sure if this comment mean you want to improve the code or if there is some bug / limitiation
It's just for improving code😂.

Reply:
I don't think there is any need for a ticket to track improving the code 👍